airbyte-integrations/connectors/source-google-sheets/google_sheets_source/google_sheets_source.py

```diff
@@ -0,0 +1,128 @@
+        config = config_container.rendered_config
+        client = Helpers.get_authenticated_sheets_client(config)
+        spreadsheet_id = config['spreadsheet_id']
+        try:
+            # Attempt to get first row of sheet
+            client.get(spreadsheetId=spreadsheet_id, includeGridData=False, ranges='1:1').execute()
+        except errors.HttpError as err:
+            reason = str(err)
+            # Give a clearer message if it's a common error like 404.
+            if err.resp.status == 404:
+                reason = "Requested spreadsheet was not found."
+
+            print(f"Formatted error: {reason}")
+            return AirbyteConnectionStatus(status=Status.FAILED, message=str(reason))
+
+        return AirbyteConnectionStatus(status=Status.SUCCEEDED)
+
+    def discover(self, logger: AirbyteLogger, config_container) -> AirbyteCatalog:
+        config = config_container.rendered_config
+        client = Helpers.get_authenticated_sheets_client(config)
+        spreadsheet_id = config['spreadsheet_id']
+        try:
+            logger.info(f"Running discovery on sheet {spreadsheet_id}")
+            spreadsheet_metadata = Spreadsheet.parse_obj(client.get(spreadsheetId=spreadsheet_id, includeGridData=False).execute())
+            sheet_names = [sheet.properties.title for sheet in spreadsheet_metadata.sheets]
+            streams = []
+            for sheet_name in sheet_names:
+                header_row_data = Helpers.get_first_row(client, spreadsheet_id, sheet_name)
+                stream = Helpers.headers_to_airbyte_stream(sheet_name, header_row_data)
+                streams.append(stream)
+            return AirbyteCatalog(streams=streams)
+
+        except errors.HttpError as err:
+            reason = str(err)
+            if err.resp.status == 404:
+                reason = "Requested spreadsheet was not found."
+            raise Exception(f"Could not run discovery: {reason}")
+
+    def read(self, logger: AirbyteLogger, config_container, catalog_path, state=None) -> Generator[AirbyteMessage, None, None]:
+        config = config_container.rendered_config
+        client = Helpers.get_authenticated_sheets_client(config)
+
+        catalog = AirbyteCatalog.parse_obj(self.read_config(catalog_path))
+        sheet_to_column_name = Helpers.parse_sheet_and_column_names_from_catalog(catalog)
+        spreadsheet_id = config['spreadsheet_id']
+
+        logger.info(f"Starting syncing spreadsheet {spreadsheet_id}")
+
+        sheet_to_column_index_to_name = Helpers.get_available_sheets_to_column_index_to_name(client, spreadsheet_id, sheet_to_column_name)
+        for sheet in sheet_to_column_index_to_name.keys():
+            logger.info(f"Syncing sheet {sheet}")
+            column_index_to_name = sheet_to_column_index_to_name[sheet]
+            row_cursor = 2  # we start syncing past the header row
+            encountered_blank_row = False
+            while not encountered_blank_row:
```

Comment:
so it's not supported to have a blank row in the middle of your data? i guess that's fine.